src/solver/optimisation/constraint_builder.h

```diff
@@ -0,0 +1,206 @@
+      case Variable::IntercoDirectCost:
+        return nativeOptimVar.NumeroDeVariableCoutOrigineVersExtremiteDeLInterconnexion[index_];
+      case Variable::IntercoIndirectCost:
+        return nativeOptimVar.NumeroDeVariableCoutExtremiteVersOrigineDeLInterconnexion[index_];
+      case Variable::ShortTermStorageInjection:
+        return nativeOptimVar.SIM_ShortTermStorage.InjectionVariable[index_];
+      case Variable::ShortTermStorageWithdrawal:
+        return nativeOptimVar.SIM_ShortTermStorage.WithdrawalVariable[index_];
+      case Variable::ShortTermStorageLevel:
+        return nativeOptimVar.SIM_ShortTermStorage.LevelVariable[index_];
+      case Variable::HydProd:
+        return nativeOptimVar.NumeroDeVariablesDeLaProdHyd[index_];
+      case Variable::HydProdDown:
+        return nativeOptimVar.NumeroDeVariablesVariationHydALaBaisse[index_];
+      case Variable::HydProdUp:
+        return nativeOptimVar.NumeroDeVariablesVariationHydALaHausse[index_];
+      case Variable::Pumping:
+        return nativeOptimVar.NumeroDeVariablesDePompage[index_];
+      case Variable::HydroLevel:
+        return nativeOptimVar.NumeroDeVariablesDeNiveau[index_];
+      case Variable::Overflow:
+        return nativeOptimVar.NumeroDeVariablesDeDebordement[index_];
+      case Variable::FinalStorage:
+        return problemeHebdo.NumeroDeVariableStockFinal[index_];
+      case Variable::LayerStorage:
+        return problemeHebdo.NumeroDeVariableDeTrancheDeStock[index_][0]; // TODO FIXMEEE
+      case Variable::PositiveUnsuppliedEnergy:
+        return nativeOptimVar.NumeroDeVariableDefaillancePositive[index_];
+      case Variable::NegativeUnsuppliedEnergy:
+        return nativeOptimVar.NumeroDeVariableDefaillanceNegative[index_];
+      default:
+        return -1;
+    }
+    }
+
+public:
+    ConstraintBuilder& include(Variable var, double coeff, int timeShift = 0, bool wrap = false)
+    {
+      std::vector<double>& Pi = problemeAResoudre.Pi;
+      std::vector<int>& Colonne = problemeAResoudre.Colonne;
+      int varIndex = getVariableIndex(var, timeShift, wrap);
+      if (varIndex >= 0)
+      {
+        Pi[nombreDeTermes_] = coeff;
+        Colonne[nombreDeTermes_] = varIndex;
+        nombreDeTermes_++;
+      }
+      return *this;
+    }
+    ConstraintBuilder& equal(double rhs)
+    {
+      operator_ = '=';
+      rhs_ = rhs;
+      return *this;
+    }
+    ConstraintBuilder& less(double rhs)
+    {
+      operator_ = '<';
+      rhs_ = rhs;
+      return *this;
+    }
+    ConstraintBuilder& greater(double rhs)
+    {
+      operator_ = '>';
+      rhs_ = rhs;
+      return *this;
+    }
+
+    void build();
+
+private:
+    PROBLEME_ANTARES_A_RESOUDRE& problemeAResoudre;
+    PROBLEME_HEBDO& problemeHebdo;
+    const std::vector<CORRESPONDANCES_DES_VARIABLES>& varNative;
+
+    unsigned int hourInWeek_ = 0;
+    // TODO: vérifier que l'index correspond bien à l'objet concerné (zone, lien, cluster thermique, etc.)
```

Comment:
Instead of the updateIndex, we could have specific classes for each kind of objects, for example:
```cpp
builder.area(1)  // Return a class specific to areas
    .include(Variable::HydProd, 1.0)  // checks that the variable is only for areas.
                                     // actually we could even have one method by acceptable variable
```

This would also enable to have nested classes for clusters, for example:
```cpp
builder.area(1)
    .cluster(3)
    .include(...)
```

Reply:
I agree, although for thermal clusters & short-term storages, the index is **global** so we don't to provide area indices.